src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,483 @@
+
+import qualified Data.Text as T
+import Data.Text.Encoding
+
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+
+import qualified Crypto.Hash.SHA256 as SHA256
+
+import Data.Aeson ( FromJSON(..), ToJSON(..), FromJSONKey(..), ToJSONKey(..)
+                  , (.:), (.:?), (.=) )
+import qualified Data.Aeson as JSON
+
+import qualified Data.SBV.Dynamic as SBV
+
+import Verifier.SAW.FiniteValue
+import Verifier.SAW.SATQuery
+import Verifier.SAW.ExternalFormat
+import Verifier.SAW.SharedTerm
+
+import SAWScript.SolverCache.LMDBOptDatabase (encodeHex, LMDBOptDatabase)
+import qualified SAWScript.SolverCache.LMDBOptDatabase as LMDBOpt
+
+import SAWScript.Options
+import SAWScript.Proof
+
+
+-- Orphan Instances and Helper Functions ---------------------------------------
+
+deriving instance Generic FirstOrderType
+deriving instance Generic FirstOrderValue
+
+-- | The options for JSON-serializing 'FirstOrderType's and 'FirstOrderValue's:
+-- remove the @FOV@/@FOT@ prefixes and encode the different constructors as
+-- two-element arrays.
+firstOrderJSONOptions :: JSON.Options
+firstOrderJSONOptions =
+  JSON.defaultOptions { JSON.sumEncoding = JSON.TwoElemArray
+                      , JSON.constructorTagModifier = dropFO }
+  where dropFO ('F':'O':tv:cs) | tv `elem` ['T', 'V'] = cs
+        dropFO cs = cs
+
+instance FromJSON FirstOrderType where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+instance FromJSON FirstOrderValue where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+
+instance ToJSON FirstOrderType where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
+instance ToJSON FirstOrderValue where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
+
+-- | Run the given IO action, but if the given 'timeout' (in ms) is reached
+-- or the action encounters any 'SomeException', 'Left' is returned
+tryWithTimeout :: Int -> IO a -> IO (Either String a)
+tryWithTimeout t_ms m = try (timeout (t_ms * 1000) m) <&> \case
+  Right (Just a) -> Right a
+  Right Nothing  -> Left $ "Operation timed out (" ++ show t_ms ++ "ms)"
+  Left (exn :: SomeException) -> Left $ show exn
+
+
+-- Solver Backends -------------------------------------------------------------
+
+-- | A datatype representing one of the solver backends available in SAW. Note
+-- that for 'SBV' and 'W4', multiple backends will be used (e.g. 'SBV' with
+-- 'Z3' or 'W4' with 'AIG' and 'ABC').
+data SolverBackend = What4
+                   | SBV
+                   | AIG
+                   | RME
+                   -- External solvers (copied from SBV.Solver)
+                   | ABC
+                   | Boolector
+                   | Bitwuzla
+                   | CVC4
+                   | CVC5
+                   | DReal
+                   | MathSAT
+                   | Yices
+                   | Z3
```

Comment:
Several of these backends are present in SBV, but not in SAW (e.g., `Bitwuzla`). I'm not sure if that makes a difference in practice, but it is a little weird seeing them here. I wonder if it's worth removing them here.

Reply:
Honestly the main reason I kept them was the convenience of having [`cvtSolver :: SBV.Solver -> SolverBackend`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverCache.hs#L183) be total. But, it also means that if anyone ever wants to add support for `Bitwuzla` or `DReal`, all they'd have to do is add a SAW command which calls `proveSBV SBV.Bitwuzla`/`SBV.DReal`. Otherwise they'd also have to update this type, the pattern match in [`cvtSolver`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverCache.hs#L183) , and the pattern match in [`getSolverBackendVersion`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverVersions.hs#L44).

That all being said, I totally understand removing them. I don't really have a preference, and am happy to do whatever you think is best.

Reply:
> But, it also means that if anyone ever wants to add support for `Bitwuzla` or `DReal`, all they'd have to do is add a SAW command which calls `proveSBV SBV.Bitwuzla`/`SBV.DReal`.

As someone who recently added support for a new SMT solver... I _wish_ it were that straightforward. :)

With the current state of SAW's code, there are quite a few places in the code that you'd need to update if you want to add a new solver. I suppose the argument here is that only listing the `SolverBackend`s that SAW currently supports would add yet another place in the code that you'd need to update, which is fair. I don't have a strong opinion on this, but if you do leave the current definition of `SolverBackend` in place, I think it would be worth putting all of the supported-in-SBV-but-not-supported-in-SAW solvers in a separate part of the definition, separated by a comment.

Reply:
> As someone who recently added support for a new SMT solver... I wish it were that straightforward. :)

Oh dang it! Shows my naivete :)

I decided to leave those extra two in, just because I think it would be very easy for someone to not realize they have to update this file if any additional SBV backends are added. But, I add a bunch of comments making it clear that SAW does not currently support those solvers, and they are only there to make sure the system for keeping track of solver backend versions doesn't silently break if support for them is ever added. How does it look now?